### PR TITLE
fix(exporter): add some parallel fetching of blocks where possible

### PR DESCRIPTION
**Pull request by dignifiedquire**
Ref https://github.com/ipfs/js-ipfs/issues/498

### CODE REVIEW

File: src/exporters/dir.js [hunk 1]
```diff
@@ -14,15 +15,17 @@ module.exports = (node, name, dagService) => {
...
+
+      cb(null, switchType(
         n,
         () => pull.values([item]),
         () => fileExporter(n, item.path, dagService)
-      )),
-      pull.flatten()
-    )),
+      ))
+    })),
```
Comment (diasdavid) [inline src/exporters/dir.js]: @dignifiedquire could you please add a bit of documentation to this function, each time I come to read it, I scratch my head to get back to understand what is supposed to be going on

Comment (dignifiedquire) [inline src/exporters/dir.js]: Is `/* this is where the magic happens */` enough?

Comment (diasdavid) [inline src/exporters/dir.js]: no. If is helpful to have diagrams, add a file to explain the internals with those graphs :)

Comment (dignifiedquire) [inline src/exporters/dir.js]: @diasdavid added a description. Can we get this merged soon plese?

File: src/exporters/dir.js [hunk 2]
```diff
@@ -2,27 +2,40 @@
...
+  // The algorithm below is as follows
+  //
+  // 1. Take all links from a given directory node
+  // 2. Map each link to their full name (parent + link name) + hash
+  // 3. Parallel map to
+  // 3.1. Resolve the hash against the dagService
+  // 3.2. Switch on the node type
+  //      - `directory`: return node
+  //      - `file`: use the fileExporter to load and return the file
+  // 4. Flatten
```
Comment (diasdavid) [inline src/exporters/dir.js]: <3

**Pull request merged**